base/reducedim.jl

```diff
@@ -321,21 +332,43 @@ julia> mapreduce(isodd, |, a, dims=1)
 """
 mapreduce(f, op, A::AbstractArrayOrBroadcasted; dims=:, init=_InitialValue()) =
     _mapreduce_dim(f, op, init, A, dims)
-mapreduce(f, op, A::AbstractArrayOrBroadcasted...; kw...) =
-    reduce(op, map(f, A...); kw...)
+
+mapreduce(f, op, A::AbstractArrayOrBroadcasted, Bs::AbstractArrayOrBroadcasted...; dims=:, init=_InitialValue()) =
+    _mapreduce_dim_vararg(f, op, init, (A, Bs...), dims)
 
 _mapreduce_dim(f, op, nt, A::AbstractArrayOrBroadcasted, ::Colon) =
     mapfoldl_impl(f, op, nt, A)
 
 _mapreduce_dim(f, op, ::_InitialValue, A::AbstractArrayOrBroadcasted, ::Colon) =
     _mapreduce(f, op, IndexStyle(A), A)
 
-_mapreduce_dim(f, op, nt, A::AbstractArrayOrBroadcasted, dims) =
-    mapreducedim!(f, op, reducedim_initarray(A, dims, nt), A)
-
 _mapreduce_dim(f, op, ::_InitialValue, A::AbstractArrayOrBroadcasted, dims) =
     mapreducedim!(f, op, reducedim_init(f, op, A, dims), A)
 
+
+_mapreduce_dim_vararg(f, op, init, As::Tuple, ::Colon) =
+    only(_mapreduce_dim_vararg(f, op, init, As, 1:maximum(ndims, As)))
+
+_mapreduce_dim_vararg(f, op, ::_InitialValue, As::Tuple, ::Colon) =
+    mapreduce(splat(f), op, zip(As...)) # fallback, without init -- needs vararg _mapreduce
+
+mapreduce_empty(f::typeof(splat(+)).name.wrapper, op, ::Type{T}) where {T<:Tuple} =
+    reduce_empty(op, Core.Compiler.return_type(op, T))  # for mapreduce(+, +, Int[], Int[])
+
+
+_mapreduce_dim_vararg(f, op, ::_InitialValue, As::Tuple, dims) =
+    _mapreduce_dim(identity, op, _InitialValue(), map(f, As...), dims) # fallback, without init -- no vararg reducedim_init
+
+_mapreduce_dim_vararg(f, op, init, As::Tuple, dims) =
+    _mapreduce_dim_vararg(f, op, init, As, dims, IteratorSize(Generator(f, As...)))
```

Comment:
The only purpose of `Generator(f, As...)` is for the traits `HasShape` etc. Perhaps there's a better way. 

Some weird test cases:
```
mapreduce(/,+,[1 2 3; 4 5 6],[7,8,9,10],dims=2,init=0.0)
mapreduce(/,+,[1 2 3; 4 5 6],[7,8,9,10],dims=1,init=0.0)
mapreduce(/,+,[1 2 3; 4 5 6],rand(2,3,1),dims=1,init=0.0)
mapreduce(/,+,rand(2,3),rand(2,3,1),dims=2,init=0.0)
```

Reply:
A bolder possibility would be to declare those to be errors. This would fit with restricting `map(f, A, B)` to `size(A)==size(B)`, instead of this `zip` behaviour. For tuples, this is still how `map` acts; in the discussion at #42216 most people seem to prefer that to the `zip`.